FreeRTOS-Plus/Demo/AWS/Fleet_Provisioning_Windows_Simulator/Fleet_Provisioning_With_CSR_Demo/mqtt_operations.c

```diff
@@ -0,0 +1,1069 @@
+                        MQTT_Status_strerror( xMqttStatus ) ) );
+        }
+        else
+        {
+            xReturnStatus = true;
+        }
+    }
+
+    return xReturnStatus;
+}
+/*-----------------------------------------------------------*/
+
+bool xUnsubscribeFromTopic( const char * pcTopicFilter,
+                            uint16_t usTopicFilterLength )
+{
+    bool xReturnStatus = false;
+    MQTTStatus_t xMqttStatus;
+    MQTTContext_t * pxMqttContext = &xMqttContext;
+    MQTTSubscribeInfo_t pxSubscriptionList[ 1 ];
+
+    configASSERT( pxMqttContext != NULL );
+    configASSERT( pcTopicFilter != NULL );
+    configASSERT( usTopicFilterLength > 0 );
+
+    /* Start with everything at 0. */
+    ( void ) memset( ( void * ) pxSubscriptionList, 0x00, sizeof( pxSubscriptionList ) );
+
+    /* This example subscribes to only one topic and uses QOS1. */
+    pxSubscriptionList[ 0 ].qos = MQTTQoS1;
+    pxSubscriptionList[ 0 ].pTopicFilter = pcTopicFilter;
+    pxSubscriptionList[ 0 ].topicFilterLength = usTopicFilterLength;
+
+    /* Generate packet identifier for the UNSUBSCRIBE packet. */
+    usGlobalUnsubscribePacketIdentifier = MQTT_GetPacketId( pxMqttContext );
+
+    /* Send UNSUBSCRIBE packet. */
+    xMqttStatus = MQTT_Unsubscribe( pxMqttContext,
+                                    pxSubscriptionList,
+                                    sizeof( pxSubscriptionList ) / sizeof( MQTTSubscribeInfo_t ),
+                                    usGlobalUnsubscribePacketIdentifier );
+
+    if( xMqttStatus != MQTTSuccess )
+    {
+        LogError( ( "Failed to send UNSUBSCRIBE packet to broker with error = %s.",
+                    MQTT_Status_strerror( xMqttStatus ) ) );
+    }
+    else
+    {
+        LogDebug( ( "UNSUBSCRIBE sent topic %.*s to broker.",
+                    usTopicFilterLength,
+                    pcTopicFilter ) );
+
+        /* Process incoming packet from the broker. Acknowledgment for unsubscribe
+         * operation ( UNSUBACK ) will be received here. This demo uses
+         * MQTT_ProcessLoop to receive packet from network. */
+        xMqttStatus = MQTT_ProcessLoop( pxMqttContext, mqttopMQTT_PROCESS_LOOP_TIMEOUT_MS );
+
+        if( xMqttStatus != MQTTSuccess )
+        {
+            LogError( ( "MQTT_ProcessLoop returned with status = %s.",
+                        MQTT_Status_strerror( xMqttStatus ) ) );
+        }
+        else
+        {
+            xReturnStatus = true;
+        }
+    }
+
+    return xReturnStatus;
+}
+/*-----------------------------------------------------------*/
+
+bool xPublishToTopic( const char * pcTopicFilter,
+                      uint16_t usTopicFilterLength,
+                      const char * pcPayload,
+                      size_t xPayloadLength )
+{
+    bool xReturnStatus = false;
+    MQTTStatus_t xMqttStatus = MQTTSuccess;
+    uint8_t ucPublishIndex = mqttopMAX_OUTGOING_PUBLISHES;
+    MQTTContext_t * pxMqttContext = &xMqttContext;
```

Comment:
This should be a parameter rather than a global.

Reply:
These mqtt_operations changes will be addressed in a future PR, where mqtt_demo_helpers is adapted to work with PKCS11.